module/os/freebsd/zfs/zfs_vnops_os.c

```diff
@@ -4449,9 +4453,87 @@ static int
 zfs_freebsd_read(struct vop_read_args *ap)
 {
 	zfs_uio_t uio;
+	int error = 0;
+	znode_t *zp = VTOZ(ap->a_vp);
+	int ioflag = ioflags(ap->a_ioflag);
+
 	zfs_uio_init(&uio, ap->a_uio);
-	return (zfs_read(VTOZ(ap->a_vp), &uio, ioflags(ap->a_ioflag),
-	    ap->a_cred));
+
+	zfs_direct_enabled_t direct =
+	    zfs_check_direct_enabled(zp, ioflag, &error);
+
+	if (direct == ZFS_DIRECT_IO_ERR) {
+		return (error);
+	} else if (direct == ZFS_DIRECT_IO_ENABLED) {
+		error =
+		    zfs_freebsd_read_direct(zp, &uio, UIO_READ, ioflag,
+		    ap->a_cred);
+		/*
+		 * XXX We occasionally get an EFAULT for Direct I/O reads on
+		 * FreeBSD 13. This still needs to be resolved. The EFAULT comes
+		 * from:
+		 * zfs_uio_get__dio_pages_alloc() ->
+		 * zfs_uio_get_dio_pages_impl() ->
+		 * zfs_uio_iov_step() ->
+		 * zfs_uio_get_user_pages().
+		 * We return EFAULT from zfs_uio_iov_step(). When a Direct I/O
+		 * read fails to map in the user pages (returning EFAULT) the
+		 * Direct I/O request is broken up into two separate IO requests
+		 * and issued separately using Direct I/O.
```

Comment:
@markjdb looked at this a long time ago. It is in the history of this PR. This was just an issue with FreeBSD 13 and there were no issues in FreeBSD 14 that caused `EFAULT` to be returned... I am not at all a FreeBSD expert, and we did this `XXX` comment in the hopes someone in the FreeBSD world would get it solved. Matt Macy did a ton of work back in the day to the FreeBSD port working. In all honesty, the FreeBSD port would not even exist if he didn't put in all that work. Unfortunately, he is no longer working on ZFS as far as I know. This patch was just to get the FreeBSD 13 CI to pass back in the day with the `direct` testing.

Would be happy to finally get a fix for this, but I am not sure how to fix it to be honest... If you are interested in where this commit came into the PR it is here: https://github.com/bwatkinson/zfs/commit/b039df893af6d74fd76c9b9fe5fec809bbf3c90c

So, unless we have someone who knows that part of the FreeBSD code well and who wants to solve this one, I am kind of at a lost on how to correctly fix it.

Reply:
I had a long conversation with @kostikbel about https://github.com/openzfs/zfs/pull/15665, where he explained to me that EFAULT is a normal scenario for FreeBSD VFS.  May be he could comment here also.  It may be that EFAULT here is not a reason to fall back to regular I/O, since one may fail the same way.  I am not ready to say whether EFAULT for direct I/O always means EFAULT for normal I/O, but in my understanding they do have some overlap.

Reply:
@kostikbel it would great if you could look at this issue and give any advice here.